alvr/launcher/src/actions.rs

```diff
@@ -137,85 +122,76 @@ pub fn get_release(
+                message: "Downloading Client APK".into(),
+                progress,
+            }));
+        })?;
         let mut file = File::create(&apk_path)?;
         file.write_all(&apk_buffer)?;
     }
 
-    worker_message_sender.send(WorkerMessage::ProgressUpdate(Progress {
-        message: "Installing APK".into(),
-        progress: 0.0,
-    }))?;
-
-    let res = match Command::new(ADB_EXECUTABLE)
-        .arg("install")
-        .arg("-d")
-        .arg(&apk_path)
-        .output()
-    {
-        Ok(res) => res,
-        Err(_) => {
-            let adb_path = data_dir().join("platform-tools").join(ADB_EXECUTABLE);
-
-            if !adb_path.exists() {
-                let mut buffer = Cursor::new(
-                    download(
-                        worker_message_sender,
-                        "Downloading Android Platform Tools",
-                        PLATFORM_TOOLS_DL_LINK,
-                        client,
-                    )
-                    .await?,
-                );
-
-                zip::ZipArchive::new(&mut buffer)?.extract(data_dir())?;
-            }
+    let layout = alvr_filesystem::Layout::new(&root);
+    let adb_path = alvr_adb::commands::require_adb(&layout, |downloaded, total| {
+        let progress = total.map(|t| downloaded as f32 / t as f32).unwrap_or(0.0);
+        let _ = worker_message_sender.send(WorkerMessage::ProgressUpdate(Progress {
+            message: "Downloading ADB".into(),
+            progress,
+        }));
+    })?;
 
-            worker_message_sender.send(WorkerMessage::ProgressUpdate(Progress {
-                message: "Installing APK".into(),
-                progress: 0.0,
-            }))?;
+    let device_serial = alvr_adb::commands::list_devices(&adb_path)?
+        .iter()
+        .find_map(|d| d.serial.clone())
+        .ok_or(anyhow::anyhow!("Failed to find connected device"))?;
 
-            Command::new(adb_path)
-                .arg("install")
-                .arg("-r")
-                .arg(&apk_path)
-                .output()?
-        }
+    let v = if release.version.starts_with('v') {
+        release.version[1..].to_string()
+    } else {
+        release.version
     };
-    if res.status.success() {
-        Ok(())
+    let version = Version::parse(&v).context("Failed to parse release version")?;
+    let stable = version.pre.is_empty() && !version.build.contains("nightly");
+    let application_id = if stable {
+        alvr_adb::PACKAGE_NAME_GITHUB_STABLE
     } else {
-        Err(anyhow::anyhow!(
-            "ADB install failed: {}",
-            String::from_utf8_lossy(&res.stderr)
-        ))
+        alvr_adb::PACKAGE_NAME_GITHUB_DEV
+    };
+
+    if alvr_adb::commands::is_package_installed(&adb_path, &device_serial, application_id)? {
+        worker_message_sender.send(WorkerMessage::ProgressUpdate(Progress {
+            message: "Uninstalling old APK".into(),
+            progress: 0.0,
+        }))?;
```

Comment:
Is it necessary to send the progress at 0%? Since that will also be called by `uninstall_package`, I'm thinking that it's not good that we have to repeat the same message two times in the code

Reply:
Ah ok, I see